FEPointEvaluation: Work towards dim < spacedim

kronbichler opened the pull request:
Fixes #11944. Works towards the problematic case discussed in https://github.com/dealii/dealii/pull/16753#issuecomment-1999354107 and the generally desired support for `dim != spacedim` in several areas of the library dependent on `FEPointEvaluation` (and eventually `FEEvaluation`, as the latter should also be supported).

@bergbauer would you mind having a look at the changes in the type traits of `FEPointEvaluation` to see if you agree with the overall direction? I did some changes in a bit of a haste, so I might have selected non-optimal typedefs.

bergbauer commented on include/deal.II/matrix_free/fe_point_evaluation.h:
```diff
@@ -187,33 +190,33 @@ namespace internal
-             const unsigned int   vector_lane,
-             const unsigned int   component)
+      static Tensor<1, spacedim, ScalarNumber>
+      access(const Tensor<1, n_components, Tensor<1, spacedim, Number>> &value,
```
```suggestion
      access(const gradient_type &value,
```

bergbauer commented on include/deal.II/matrix_free/fe_point_evaluation.h:
```diff
@@ -223,10 +226,12 @@ namespace internal
       using value_type               = Number;
       using scalar_value_type        = ScalarNumber;
       using vectorized_value_type    = VectorizedArrayType;
-      using gradient_type            = Tensor<1, dim, Number>;
+      using unit_gradient_type       = Tensor<1, dim, Number>;
+      using gradient_type            = Tensor<1, spacedim, Number>;
       using scalar_gradient_type     = Tensor<1, dim, ScalarNumber>;
       using vectorized_gradient_type = Tensor<1, dim, VectorizedArrayType>;
-      using interface_vectorized_gradient_type = vectorized_gradient_type;
+      using interface_vectorized_gradient_type =
+        Tensor<1, dim, VectorizedArrayType>;
```
same here with `*_unit_*`

bergbauer commented on include/deal.II/matrix_free/fe_point_evaluation.h:
```diff
@@ -341,31 +346,31 @@ namespace internal
       static void
-      access(gradient_type     &value,
-             const unsigned int vector_lane,
+      access(Tensor<1, spacedim, Number> &value,
```
```suggestion
      access(gradient_type &value,
```

bergbauer commented on include/deal.II/matrix_free/fe_point_evaluation.h:
```diff
@@ -375,6 +380,7 @@ namespace internal
       using value_type               = Tensor<1, dim, Number>;
       using scalar_value_type        = Tensor<1, dim, ScalarNumber>;
       using vectorized_value_type    = Tensor<1, dim, VectorizedArrayType>;
+      using unit_gradient_type       = Tensor<2, dim, Number>;
```
I would remove this here as this is a special case `dim = spacedim`.

kronbichler replied:
I need to check this more.

bergbauer commented on include/deal.II/matrix_free/fe_point_evaluation.h:
```diff
@@ -341,31 +346,31 @@ namespace internal
-      access(const gradient_type &value,
-             const unsigned int   vector_lane,
+      static Tensor<1, spacedim, ScalarNumber>
+      access(const Tensor<1, spacedim, Number> &value,
```
```suggestion
      access(const gradient_type &value,
```

bergbauer commented on include/deal.II/matrix_free/fe_point_evaluation.h:
```diff
@@ -341,31 +346,31 @@ namespace internal
       }
 
       static void
-      access(gradient_type     &value,
-             const unsigned int vector_lane,
+      access(Tensor<1, spacedim, Number> &value,
+             const unsigned int           vector_lane,
              const unsigned int,
-             const scalar_gradient_type &shape_gradient)
+             const Tensor<1, spacedim, ScalarNumber> &shape_gradient)
```
This should stay `scalar_gradient_type` (with `scalar_gradient_type` adjusted).

kronbichler replied:
Can you elaborate here? I tried to play with this parameter and it gets used in both `get_gradient`/`set_gradient` where I need it to be a tensor of entries, and in the interface you are pointing two. But that latter interfaces with the slow path and `FEValues`, where I believe the spacedim variant is fine. I might be missing something here, so I would appreciate a further comment.

bergbauer commented on include/deal.II/matrix_free/fe_point_evaluation.h:
```diff
@@ -264,7 +269,7 @@ namespace internal
                    const scalar_gradient_type &result)
       {
-        for (unsigned int d = 0; d < dim; ++d)
+        for (unsigned int d = 0; d < spacedim; ++d)
```
This looks wrong `scalar_gradient_type`  is `dim` long and you loop until `spacedim`?

bergbauer commented on include/deal.II/matrix_free/fe_point_evaluation.h:
```diff
@@ -2639,13 +2648,13 @@ FEPointEvaluation<n_components_, dim, spacedim, Number>::evaluate_fast(
-                  apply_diagonal_transformation(this->inverse_jacobian_ptr[0],
-                                                unit_gradient) :
+                  apply_diagonal_transformation(
+                    this->inverse_jacobian_ptr[0].transpose(), unit_gradient) :
```
Are you sure about this? Why is the transpose necessary here?

kronbichler replied:
We need to transpose here to have consistent dimensions for the function (dim/spacedim), even though we only ever access the diagonal. In terms of correctness, it should not matter, but I also think it is more readable because also the `else` case in this check uses the transposed Jacobian. My assumption was that the transposed matrix is never built in release mode, because the entries are unused.

bergbauer replied:
Then we also need it for the `do_evaluate_in_face` function. Can you add a test for faces?

I am not sure about the transposition. I thought about an `apply_transposed_transformation` function in the past to make this explicit. We should check this in the machine code at some point, but I think it is not necessary now.

bergbauer commented on include/deal.II/non_matching/mapping_info.h:
```diff
@@ -1960,8 +1960,8 @@ namespace NonMatching
+                    for (unsigned int s = 0; s < dim; ++s)
                       dealii::internal::VectorizedArrayTrait<Number>::get(
                         inverse_jacobians[is_interior ? 0 : 1]
                                          [compressed_offset][s][d],
```
Can't we simply transpose indices here?

bergbauer reviewed (commented):
Nice step forward!

kronbichler commented on include/deal.II/matrix_free/fe_point_evaluation.h:
```diff
@@ -68,7 +68,10 @@ namespace internal
+        Tensor<1, n_components, Tensor<1, spacedim, Number>>;
       using scalar_gradient_type =
         Tensor<1, n_components, Tensor<1, dim, ScalarNumber>>;
       using vectorized_gradient_type =
```
I opted to keep the names without `_unit` for the parts involving `dim` and choose `real_gradient_type` for the space-dim gradient. I admit that in the user interface I have to define `FEPointEvaluation::gradient_type` to be an alias for `real_gradient_type`, but I think all the names are in the internal namespace and most compact this way. One could interpret them as being the user-facing interface in `spacedim` and the evaluator-facing interface in `dim` (which is exactly what we are doing). What do you think?

bergbauer replied:
Honestly, I think it would be better to be explicit and call every `dim` gradient `_unit_` and every `spacedim` gradient `_real_`. Even if it is a bit lengthier, it forces us to be consistent.

kronbichler commented on include/deal.II/matrix_free/fe_point_evaluation.h:
```diff
@@ -563,164 +590,6 @@ namespace internal
       }
     };
 
-    template <typename Number>
-    struct EvaluatorTypeTraits<1, 1, Number>
-    {
-      using ScalarNumber =
```
Note that I removed this partial specialization, because with `spacedim` being another parameter, I would have needed even more cases. Instead, I now added an `std::enable_if_t` to the specialization with `n_components == dim` that further checks `dim > 1` to avoid ambiguous specializations for `n_components = 1` and `n_components = dim`. I think this gives us cleaner code.

kronbichler commented:
@bergbauer I addressed your comments. The work needed to get multiple components and `FEFacePointEvaluation` running was more involved than I anticipated, but we now have tests for those and the code should run and compile. We might have some hidden features not completely covered yet for `dim < spacedim`, but I think this PR does a lot to extend the infrastructure.

bangerth commented:
The test `matrix_free/is_supported_01` is failing.

bergbauer commented on tests/matrix_free/point_evaluation_32.cc:
```diff
@@ -0,0 +1,167 @@
+                             EvaluationFlags::values |
+                               EvaluationFlags::gradients);
+          deallog << "face number " << f << std::endl;
+          for (unsigned int i = 0; i < unit_points.size(); ++i)
+            deallog << "value " << evaluator.get_value(i) << " gradient "
+                    << evaluator.get_gradient(i) << std::endl;
+          deallog << std::endl;
```
Should we also compare to `FEFaceValues` like in the other tests? How did you check if the results are correct?

kronbichler replied:
I added `FEFaceValues` now; we didn't have it in the test I copied it from, but there is no damage in adding it. We now also print the reference and see that the values are OK. I had previously just verified the part with linear function in dim=1, spacedim=1 with a rectangular domain, but of course it is good to verify the data on the sphere as well.

kronbichler commented:
Comments are fixed, ready for re-review.

bergbauer reviewed (commented):
This looks great now, nice work! @kronbichler  Can somebody else have second look, maybe @peterrum ?

Is this already used in the `VectorTools` functions etc now due to the changes in `is_supported`?

peterrum commented on tests/matrix_free/is_supported_01.output:
```diff
@@ -4,20 +4,20 @@ DEAL::FE_BDM<2>(1) supported by MatrixFree: false
 DEAL::FE_DGBDM<2>(1) supported by MatrixFree: false
 DEAL::FE_DGNedelec<2>(1) supported by MatrixFree: false
-DEAL::FE_DGP<1,2>(1) supported by MatrixFree: false
+DEAL::FE_DGP<1,2>(1) supported by MatrixFree: true
```
Does this has any side effects?

See also the comment by @bergbauer 

> Is this already used in the VectorTools functions etc now due to the changes in is_supported?

kronbichler replied:
It should not, because we don't allow for compiling/linking with `DoFHandler<dim, spacedim>` when `spacedim != dim`. So I believe the user gets an error early on. I have a slight hope to be able to tackle #9544 before the release, when this comes more naturally.
